File: packages/less/src/less/tree/debug-info.js
/**
 * @deprecated The dumpLineNumbers option is deprecated. Use sourcemaps instead.
 * This will be removed in a future version.
 * 
 * @param {Object} ctx - Context object with debugInfo
 * @returns {string} Debug info as CSS comment
 */
function asComment(ctx) {
    return `/* line ${ctx.debugInfo.lineNumber}, ${ctx.debugInfo.fileName} */\n`;
}

/**
 * @deprecated The dumpLineNumbers option is deprecated. Use sourcemaps instead.
 * This function generates Sass-compatible debug info using @media -sass-debug-info syntax.
 * This format had short-lived usage and is no longer recommended.
 * This will be removed in a future version.
 * 
 * @param {Object} ctx - Context object with debugInfo
 * @returns {string} Sass-compatible debug info as @media query
 */
function asMediaQuery(ctx) {
    let filenameWithProtocol = ctx.debugInfo.fileName;
    if (!/^[a-z]+:\/\//i.test(filenameWithProtocol)) {
        filenameWithProtocol = `file://${filenameWithProtocol}`;
    }
    return `@media -sass-debug-info{filename{font-family:${filenameWithProtocol.replace(/([.:/\\])/g, function (a) {
        if (a == '\\') {
            a = '/';
        }
        return `\\${a}`;
    })}}line{font-family:\\00003${ctx.debugInfo.lineNumber}}}\n`;
}

/**
 * Generates debug information (line numbers) for CSS output.
 * 
 * @param {Object} context - Context object with dumpLineNumbers option
 * @param {Object} ctx - Context object with debugInfo
 * @param {string} [lineSeparator] - Separator between comment and media query (for 'all' mode)
 * @returns {string} Debug info string
 * 
 * @deprecated The dumpLineNumbers option is deprecated. Use sourcemaps instead.
 * All modes ('comments', 'mediaquery', 'all') are deprecated and will be removed in a future version.
 * The 'mediaquery' and 'all' modes generate Sass-compatible @media -sass-debug-info output
 * which had short-lived usage and is no longer recommended.
 */
function debugInfo(context, ctx, lineSeparator) {
    let result = '';
    if (context.dumpLineNumbers && !context.compress) {
        switch (context.dumpLineNumbers) {
            case 'comments':
                result = asComment(ctx);
                break;
            case 'mediaquery':
                result = asMediaQuery(ctx);
                break;
            case 'all':
                result = asComment(ctx) + (lineSeparator || '') + asMediaQuery(ctx);
                break;
        }
    }
    return result;
}

export default debugInfo;

File: packages/less/src/less/tree/debug-info.go
package tree

import (
	"fmt"
	"regexp"
	"strings"
)

// DebugContext represents the context containing debug information
type DebugContext struct {
	DebugInfo struct {
		LineNumber interface{}
		FileName   string
	}
}

// Context represents the compilation context
type Context struct {
	DumpLineNumbers string
	Compress        bool
}

// asComment generates a comment string with line number and filename
func asComment(ctx *DebugContext) string {
	return fmt.Sprintf("/* line %v, %s */\n", ctx.DebugInfo.LineNumber, ctx.DebugInfo.FileName)
}

// asMediaQuery generates a media query string with line number and filename
func asMediaQuery(ctx *DebugContext) string {
	filenameWithProtocol := ctx.DebugInfo.FileName
	if !regexp.MustCompile(`^[a-z]+://`).MatchString(strings.ToLower(filenameWithProtocol)) {
		filenameWithProtocol = "file://" + filenameWithProtocol
	}

	// Replace special characters with escaped versions
	escapedFilename := strings.ReplaceAll(filenameWithProtocol, "\\", "/")
	escapedFilename = regexp.MustCompile(`([.:/\\])`).ReplaceAllString(escapedFilename, "\\$1")

	// Format line number to match JavaScript format exactly
	lineNumber := fmt.Sprintf("%d", ctx.DebugInfo.LineNumber)
	return fmt.Sprintf("@media -sass-debug-info{filename{font-family:%s}line{font-family:\\00003%s}}\n",
		escapedFilename, lineNumber)
}

// DebugInfo generates debug information based on the context and debug context
func DebugInfo(context *Context, ctx *DebugContext, lineSeparator string) string {
	if context.DumpLineNumbers == "" || context.Compress {
		return ""
	}

	var result string
	switch context.DumpLineNumbers {
	case "comments":
		result = asComment(ctx)
	case "mediaquery":
		result = asMediaQuery(ctx)
	case "all":
		result = asComment(ctx)
		if lineSeparator != "" {
			result += lineSeparator
		}
		result += asMediaQuery(ctx)
	}
	return result
} 